### ISSUE TITLE
Crash on landing, probably caused by train snapping

### ISSUE THREAD

**Issue by mvolfik** [Kiplacon/RenaiTransportation#3]
Hello, there is a crash happening to me when flying trains land in various situations. Here is the traceback:
```
9485.544 Error MainLoop.cpp:1207: Exception at tick 189292: The mod Renai Transportation (0.4.6) caused a non-recoverable error.
Please report this error to the mod author.

Error while running event RenaiTransportation::on_nth_tick(1)
Trying to change direction of automatic train moving on a path.
stack traceback:
	__RenaiTransportation__/control.lua:1104: in function <__RenaiTransportation__/control.lua:773>
stack traceback:
	[C]: in function '__newindex'
	__RenaiTransportation__/control.lua:1104: in function <__RenaiTransportation__/control.lua:773>
```
It happened for example when a very slow train lands after a very short jump, or when multiple trains "overlapping" in air land.

After looking into the code of the mod, my guess on what happens is the following: there is a moving train on ground (the not-yet-airborne part of the slow train, or the first of the overlapping trains); now some part of some train "lands" - which the mod does by placing it, as a player, however, that results in the train snapping to the existing one. And now when the mod tries to change some properties, BOOM. __`Trying to change direction of an automatic train moving on path.`__

---

_Otherwise excellent work with the mod! The trains are awesome. Do you think you could implement classic tunnels for trains with the knowledge you have from here? (If it's not too conservative for you :smile:)_

**Comment by Kiplacon** [comment]
Hi mvllofik, thanks for the report and i'm glad to hear you're having fun with the train ramps 😎. A regular train bridge/tunnel is definitely something I'd like to make and it all makes sense in my head code-wise except for how it would do autopathing and two trains coming from both sides at the same time.

I didn't even know something like that was a crash condition, you taught me something.
Though I don't really understand the situation you're talking about that caused the crash, do you have a screenshot/video/MS Paint scribble that might help me visualize it?

**Comment by mvolfik** [comment]
I'm unfortunately away from PC until Friday evening, but @prokoprandacek played with me, so he might be available to create a minimal reproducible example savefile without any other mods.

Here is a simple sketch created with instagram :))

![IMG_20200902_094227_235.jpg](https://user-images.githubusercontent.com/31281386/91950520-b1621700-ed00-11ea-83ae-ec57f91ebc60.jpg)

**Comment by Kiplacon** [comment]
lmao I'm liking that diagram. idk when that would ever happen in regular play but if at the very least I could make it not crash I'll see that as an absolute win

**Comment by ProkopRandacek** [comment]
Hi!
[Video demonstation](https://streamable.com/vonld4)
[Safe file](https://randacek.dev/tmp/factorio_bug_report.zip)

PS we are using my customized version without the train station skipping so the line numbers are probably off.

**Comment by mvolfik** [comment]
@Kiplacon you mean the traceback? That is in the first comment of this issue

**Comment by mvolfik** [comment]
Oh yeah, sorry. The numbers in the video are incorrect, but I reproduced it and reported the traceback above with the latest unmodified version from mod portal on Tuesday, so it is the line 1104 in the current source.

**Comment by mvolfik** [comment]
The interesting thing is, that I think I've always seen it with the line 1104, never with 1102 (the other speed assignment when the `if` is true). So it might have something to do with how trains are defaultly oriented (until you rotate them manually) when you place them on rails. So maybe just assigning speed before setting the train path might be all that's needed. (However, if that was the exact cause, if would cause trouble in more cases than just the weird ones we got here.)

If you don't solve it until Friday, I'll try to fiddle with it more during the weekend to try to narrow the possible root causes

**Comment by Kiplacon** [comment]
I figured it out, when a train with automatic driving lands and doesn't have a route to its destination, it immediately stops to 0 speed. When another train was, like you were thinking, trying to reconnect with that stationary train, it was crashing because I only had the code check for if the train had positive speed or NOT positive speed. Once I added the distinction to check for positive speed, negative speed, or 0 speed, the crash didn't happen anymore. 

I also found another bug trying to recreate the crash, it turns out the prop cars im using to guide the train animation could be destroyed by nearby train explosions causing another crash so I fixed that too, so thanks x2 😄 . I'll upload the fix tomorrow because i'm almost done with another "interesting" feature I'll probably have done either today or tomorrow. 

btw in recent versions the train ramp can be swapped between a red mode that works the way it does now, and a blue mode that doesn't do the station skipping. hover and press Interact to swap.

Fix proof
https://i.imgur.com/rZvQQ7S.mp4

literally one line of text to fix lmao, so if you still want to use your own version, i think this should do it
![image](https://user-images.githubusercontent.com/12401534/92182400-a222bc80-ee19-11ea-9e46-4c6c8b7a5d80.png)

also if you want to add the invisible guide car fix, it just this
![image](https://user-images.githubusercontent.com/12401534/92182969-04c88800-ee1b-11ea-9b16-45744bc80719.png)
